src/main/scala/org/deepdive/ddlog/DeepDiveLogSemanticChecker.scala

```diff
@@ -142,6 +147,43 @@ object DeepDiveLogSemanticChecker extends DeepDiveLogHandler {
     }
   }
 
+  // check if head variables have bindings in the body
+  def checkHeadVariables(stmt: Statement) {
+    var variableSet = new HashSet[String]()
+    def addVariable(body: List[Body]) {
```

Comment:
I'd suggest defining a pure function of type `Expr => Set[String]` as well as `Patt => Set[String]` that return set of variables used in `Expr` and those defined in `Patt`.  Each can have `Body => Set[String]` built on top of them.  Later, any non-empty used - def set is an error using undefined variable, and non-empty def - used can be a warning for defining unused variable.  Collecting free/bound variables can be a useful utility for other purposes, so I think it's better to have them written well in a reusable form.